src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c

```diff
@@ -0,0 +1,1557 @@
+    if (NULL == read_data)
+    {
+        return -1;
+    }
+
+    read_data->direction = GRPC_C_EVENT_DIRECTION_INCOMING;
+    grpc_metadata_batch * metadata_batch = NULL;
+    struct grpc_c_metadata_t * metadata = NULL;
+    u32 key = 0;
+    u32 initial_metadata_buffer_offset = 0;
+    u32 trailing_metadata_buffer_offset = 0;
+    u32 offset = 0;
+    grpc_chttp2_stream * stream_ptr = NULL;
+    grpc_chttp2_transport * transport_ptr = NULL;
+
+    u32 pid = bpf_get_current_pid_tgid() >> 32;
+    u32 fd = 0;
+    u64 version = lookup_version(pid);
+    if (GRPC_C_VERSION_UNSUPPORTED == version)
+    {
+        return 0;
+    }
+    read_data->timestamp = bpf_ktime_get_ns();
+    transport_ptr = (grpc_chttp2_transport *)PT_REGS_PARM1(ctx);
+    stream_ptr = (grpc_chttp2_stream *)PT_REGS_PARM2(ctx);
+    read_data->stack_id = socket_stack_traces.get_stackid(
+        ctx, BPF_F_USER_STACK | BPF_F_REUSE_STACKID);
+
+    if (NULL == transport_ptr || NULL == stream_ptr)
+    {
+        return -1;
+    }
+
+    if (0 != get_stream_id((grpc_chttp2_stream *)stream_ptr, &read_data->stream_id, version))
+    {
+        return -1;
+    }
+
+    if (0 != get_fd_from_transport((grpc_chttp2_transport *)transport_ptr, &fd))
+    {
+        return -1;
+    }
+
+    struct conn_info_t* conn_info = get_or_create_conn_info(pid, fd);
+    if (NULL == conn_info)
+    {
+        return -1;
+    }
+    read_data->conn_id = conn_info->conn_id;
+
+    switch (version)
+    {
+        case GRPC_C_V1_19_0:
+            initial_metadata_buffer_offset = 0x1e0;
+            trailing_metadata_buffer_offset = 0x2d8;
+            break;
+        case GRPC_C_V1_24_1:
+            // Not supported yet.
+            break;
+        case GRPC_C_V1_33_2:
+            initial_metadata_buffer_offset = 0x330;
+            trailing_metadata_buffer_offset = 0x570;
+            break;
+        case GRPC_C_V1_41_1:
+            // Not supported yet.
+            break;
+        default:
+            return -1;
+    }
+
+    if (is_initial)
+    {
+        offset = initial_metadata_buffer_offset;
+    }
+    else
+    {
+        offset = trailing_metadata_buffer_offset;
+    }
+
+    if (0 == offset)
+    {
+        // Offset unknown for this version.
+        return 0;
+    }
+
+    metadata_batch = (grpc_chttp2_stream *)stream_ptr + offset;
+    if (NULL == metadata_batch)
+    {
+        return 0;
+    }
+
+    metadata = initiate_empty_grpc_metadata();
+    if (NULL == metadata)
+    {
+        return -1;
+    }
+
+    if (0 != fill_metadata_from_mdelem_list((grpc_mdelem_list *)metadata_batch, metadata))
+    {
+        return -1;
+    }
+
+    if (0 != fire_metadata_events(
+        metadata,
+        read_data->conn_id,
+        read_data->stream_id,
+        read_data->timestamp,
+        read_data->stack_id,
+        read_data->direction,
+        ctx))
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * @brief   Handle data being received.
```

Comment:
Combining this with another comment of yours, I thought it might even be better to add a link to the original function in github.
So I added links instead of the signatures. If you'd like signatures too, let me know.